Introduce environment checker utility

nhoriguchi opened the pull request:
This pull request is suggesting to introduce a tool to check whether required libraries and settings are properly added. According to [the document](https://software.intel.com/en-us/articles/properly-detecting-intel-software-guard-extensions-in-your-applications), we can use `sgx_cap_get_status()` to detect SGX availability, so this tool does it. I expect that it would help users find setting issues for their own.

I show below how `make prerequisite-check` works as examples:
- when the system supports and enables SGX:
  ~~~
  ubuntu@sgx:~/minbft$ make prerequisite-check
  OS: Ubuntu 18.04.3 LTS
  Go version: 1.13.6
  Intel SGX SDK version: 2.8.100.3
  Intel SGX is supported and enabled on this system.
  Intel SGX PSW version: 2.8.100.3-bionic1
  Enviroment check passed.
  ~~~
- when SGX is unavailable:
  ~~~
  ubuntu@nosgx:~/minbft$ make prerequisite-check
  OS: Ubuntu 18.04.3 LTS
  Go version: 1.13.6
  Intel SGX SDK version: 2.3.101.46683
  This CPU does not support Intel SGX.
  Makefile:8: recipe for target 'check' failed
  make[1]: *** [check] Error 1
  Intel SGX is not available. Please set environment variable SGX_MODE=SIM.
  Makefile:82: recipe for target 'prerequisite-check' failed
  make: *** [prerequisite-check] Error 1
  ~~~
- when SGX is unavailable and `SGX_MODE=SIM` is set:
  ~~~
  ubuntu@nosgx:~/minbft$ SGX_MODE=SIM make prerequisite-check
  OS: Ubuntu 18.04.3 LTS
  Go version: 1.13.6
  Intel SGX SDK version: 2.3.101.46683
  Enviroment check passed.
  ~~~

Expected to resolve #150.

Some notes which we might want to discuss:
- I mainly borrowed code from [`sgx_enable.c`](https://github.com/intel/sgx-software-enable/blob/master/sgx_enable.c) for `check-sgx-capability.c`, but substantially removed lines because it does "software-enable" part not only detecting SGX support. In this case, how do we handle license matter?
- Currently `prerequisite-check.sh` aborts if version check fails, but that might be too strict. If you think just warning is fine, I'll do it. And the script also has an assumption of using Ubuntu, that might be undesirable limitation in a long run.

nhoriguchi committed:
Makefile: Make target build depend on prerequisite-check

This enforces users to run make prerequisite-check before building,
that would help them avoid deviating from intended usage.

SKIP_PREREQUISITE_CHECK is defined for expert to skip the check.

Signed-off-by: Naoya Horiguchi <naoya.horiguchi@nec.com>

ynamiki commented:
> I mainly borrowed code from sgx_enable.c for check-sgx-capability.c, but substantially removed lines because it does "software-enable" part not only detecting SGX support. In this case, how do we handle license matter?

Original code is [licensed](https://github.com/intel/sgx-software-enable#license) under [BSD 3-Clause](https://opensource.org/licenses/BSD-3-Clause). It is [said](https://softwareengineering.stackexchange.com/questions/40561/is-bsd-license-compatible-with-apache) that BSD 3-Clause is compatible with Apache 2.0 (our license). I think what we should do is only copying (keeping) the comments about copyright and license on the top of the original file (this is requirement stated in the license). That is, you have modified a BSD-licensed file and incorporated it into our Apache-licensed project.

Note that, I am not a lawyer.

nhoriguchi commented:
> > I mainly borrowed code from sgx_enable.c for check-sgx-capability.c, but substantially removed lines because it does "software-enable" part not only detecting SGX support. In this case, how do we handle license matter?
> 
> Original code is [licensed](https://github.com/intel/sgx-software-enable#license) under [BSD 3-Clause](https://opensource.org/licenses/BSD-3-Clause). It is [said](https://softwareengineering.stackexchange.com/questions/40561/is-bsd-license-compatible-with-apache) that BSD 3-Clause is compatible with Apache 2.0 (our license). I think what we should do is only copying (keeping) the comments about copyright and license on the top of the original file (this is requirement stated in the license). That is, you have modified a BSD-licensed file and incorporated it into our Apache-licensed project.

I updated the license header of the file, thank you!

sergefdrv commented on tools/sgx-capability/check-sgx-capability.c:
```diff
@@ -0,0 +1,106 @@
+ */
+
+/*
+ * Copyright (C) 2011-2019 Intel Corporation. All rights reserved.
```
Maybe also mention where the BSD 3-Clause code comes from?

nhoriguchi replied:
Ok, I'll mention it.

sergefdrv commented on tools/sgx-capability/check-sgx-capability.c:
```diff
@@ -0,0 +1,106 @@
+		printf("A UEFI boot is required to determine whether or not your BIOS\n");
+		printf("supports Intel SGX, and to attempt the software enable.\n");
+		printf("Alternatively, you can check your BIOS to see if it has an\n");
+		printf("explicit option to enable SGX.\n");
```
Maybe we should not mention software enable in the output as it might be a bit confusing? I suggest the following output text:

```
This processor supports Intel SGX but was booted in legacy mode.
A UEFI boot is required to determine whether or not your BIOS
supports Intel SGX.
```

nhoriguchi replied:
Thanks, I'll update this message.

sergefdrv commented on tools/prerequisite-check.sh:
```diff
@@ -0,0 +1,104 @@
+	echo "Environment variable SGX_SDK not set, please make sure to source the environment file given by SGX SDK."
+	exit 1
+fi
+
```
Maybe output something here if SGX simulation mode is used?

nhoriguchi replied:
Sounds helpful. I'll add it.

sergefdrv reviewed (approved):
Thank you, I think this is a great improvement in building experience.

ynamiki commented on tools/prerequisite-check.sh:
```diff
@@ -0,0 +1,104 @@
+		return 0
+	fi
+
+	dpkg -l libsgx-enclave-common > /dev/null
```
I think it is too strict about the assumption of Ubuntu, because Intel SGX [officially supports](https://01.org/intel-softwareguard-extensions/downloads/intel-sgx-linux-2.3.1-release) not only Ubuntu but also Red Hat Enterprise Linux, CentOS, Fedora, and SUSE Linux Enterprise Server, and our code don't depends on specific operating system.

How about checking existence of `/opt/intel/sgxpsw/uninstall.sh` here not to depend on OS-specific package managers? According to the document ([v2.3.1](https://download.01.org/intel-sgx/linux-2.3.1/docs/Intel_SGX_Installation_Guide_Linux_2.3.1_Open_Source.pdf) and [v2.8](https://download.01.org/intel-sgx/sgx-linux/2.8/docs/Intel_SGX_Installation_Guide_Linux_2.8_Open_Source.pdf) (the latest release), p. 12), Intel SGX PSW installs the script in the path.

nhoriguchi replied:
> How about checking existence of `/opt/intel/sgxpsw/uninstall.sh` here not to depend on OS-specific package managers? According to the document ([v2.3.1](https://download.01.org/intel-sgx/linux-2.3.1/docs/Intel_SGX_Installation_Guide_Linux_2.3.1_Open_Source.pdf) and [v2.8](https://download.01.org/intel-sgx/sgx-linux/2.8/docs/Intel_SGX_Installation_Guide_Linux_2.8_Open_Source.pdf) (the latest release), p. 12), Intel SGX PSW installs the script in the path.

It seems that `/opt/intel/sgxpsw/uninstall.sh` is not installed when installing SGX PSW from Debian repo, so we can't rely on this file as generic solution. Actually, the installation guide mentions the installer `sgx_linux_<os>_x64_psw_<version>.bin` in "Uninstall the Intel®SGX PSW package" section, but that file does not exist, so this paragraph might be outdated.

I thought that we could use `/usr/lib/x86_64-linux-gnu/libsgx_enclave_common.so.1` as a indicator, but that's located differently depending on Linux distributions (for example it's `/usr/lib64/libsgx_enclave_common.so.1` for RHEL). Another thought is to match "libsgx_enclave_common.so" over `find /usr/` or `ldconfig -p` output, although it might look like a hack.

ynamiki replied:
Thanks for the info, I have no good idea and agree with this should not be complicated process; it is no problem with proceeding with this implementation.

sergefdrv replied:
What configure scripts usually do to detect missing dependencies is try building a tiny program (usually inlined in the script) and check if the dependency is resolved. But this may be too complicated for us at this stage.

nhoriguchi committed:
Introduce check-sgx-capability tool

Determining that Intel SGX is supported and enabled in the platform is
an important and non-trivial task. So this patch is suggesting to
introduce a helper tool for it. It should make users feel easier to get
started, and might make troubleshooting a little easier.

Signed-off-by: Naoya Horiguchi <naoya.horiguchi@nec.com>

nhoriguchi committed:
Introduce tools/prerequisite-check.sh

There're some requirements for SGX related libraries/settings, which differs
a little depending on SGX_MODE, so introducing a guiding script would be
helpful for users to know if something is wrong.

Signed-off-by: Naoya Horiguchi <naoya.horiguchi@nec.com>

ynamiki commented on Makefile:
```diff
@@ -78,5 +78,8 @@ lint:
 	go generate ./...
 
+prerequisite-check:
+	@bash tools/prerequisite-check.sh
```
Can we add a path to skip this check like `make PREREQUISITE_CHECK=0` or another way? It is useful for an expert who wants to run this program not on a recommended environment.

nhoriguchi replied:
Sure, thanks for the suggestion. I'll define `SKIP_PREREQUISITE_CHECK` to do prerequisite check by default.

nhoriguchi commented:
Summary of changes:
- improved output messages.
- `check-sgx-capability.c`: added reference to original source file,
- `prerequisite-check.sh`: renamed `check_sgx_psw_installed` with `check_sgxpsw_version` (for consistency with `check_sgxsdk_version`)
- introduced `get_sgx_psw_version` (with adding some support for RHEL/CentOS, but we still have no generic solution,)

sergefdrv commented on tools/prerequisite-check.sh:
```diff
@@ -0,0 +1,131 @@
+	exit 1
+fi
+
+if [ "$SGX_MODE" == SIM ] ; then
```
```suggestion
if [ "$SGX_MODE" != HW ] ; then
```

cf. https://github.com/hyperledger-labs/minbft/blob/335b81417ed63a3ebc2ef9805407fc58e8afeb98/usig/sgx/Makefile#L103.

sergefdrv replied:
Sorry that I missed that in the last review round, but there are a couple more such comparisons. Namely, in `check_sgx_support` and `check_sgxpsw_version`.

sergefdrv replied:
We could do the other way around and always build in hardware mode unless `"$SGX_MODE" == "SIM"`. But this would not match the convention given by SGX SDK...

nhoriguchi replied:
Thanks for the input. I found that if we use `"$SGX_MODE" != "HW"`, default case (leaving `$SGX_MODE` empty) is considered as simulation mode, which seems inconsistent with `usig/sgx/Makefile` (which sets `SGX_MODE ?= HW`, so running in hardware mode by default). So I'll fix it by giving default value of `SGX_MODE` inside the script.

> We could do the other way around and always build in hardware mode unless `"$SGX_MODE" == "SIM"`. But this would not match the convention given by SGX SDK...

I'm OK to use simulation mode If `SGX_MODE` is set to some string other than "", "HW" and "SIM". The check script will show the running mode in such a case.

sergefdrv reviewed (changes requested):
Looks very good, just a small comment.

nhoriguchi committed:
Update tools/prerequisite-check.sh

- SGX_MODE="": hardware mode
- SGX_MODE="HW": hardware mode
- SGX_MODE="SIM": simulation mode
- SGX_MODE="<any_other_string>": simulation mode

Signed-off-by: Naoya Horiguchi <naoya.horiguchi@nec.com>
Co-Authored-By: Sergey Fedorov <serge.fdrv@gmail.com>
